core/src/main/scala/com/gu/tableversions/core/InMemoryTableVersions.scala

```diff
@@ -27,39 +42,62 @@ class InMemoryTableVersions[F[_]] private (allUpdates: Ref[F, TableUpdates])(imp
       tableUpdates <- allTableUpdates
         .get(table)
         .fold(F.raiseError[List[TableUpdate]](new Exception(s"Table '$table' not found")))(F.pure)
-      operations = tableUpdates.flatMap(_.partitionUpdates)
-    } yield InMemoryTableVersions.applyUpdate(TableVersion.empty)(operations)
-
-  override def nextVersions(table: TableName, partitions: List[Partition]): F[Map[Partition, Version]] = {
-    def maxVersions(operations: List[(Partition, Version)]): Map[Partition, Version] =
-      operations.foldLeft(Map.empty[Partition, Version]) {
-        case (agg, (partition, partitionVersion)) =>
-          val previousVersion =
-            agg.getOrElse(partition, Version(0))
-          val maxVersion =
-            if (previousVersion.number > partitionVersion.number) previousVersion else partitionVersion
-
-          agg + (partition -> maxVersion)
-      }
-
-    def nextVersion(previousVersion: Option[Version]): Version =
-      previousVersion
-        .map(v => Version(v.number + 1))
-        .getOrElse(Version(1))
-
-    for {
-      allTableUpdates <- allUpdates.get
-      tableUpdates <- allTableUpdates
-        .get(table)
-        .fold(F.raiseError[List[TableUpdate]](new Exception(s"Table '${table.fullyQualifiedName}' not found")))(F.pure)
-      addedPartitions = tableUpdates.flatMap(_.partitionUpdates).collect {
-        case AddPartitionVersion(partition, version) => (partition, version)
-      }
-      maxUsedVersions = maxVersions(addedPartitions)
-      nextVersions = partitions.map(p => p -> nextVersion(maxUsedVersions.get(p)))
-    } yield nextVersions.toMap
+      operations = tableUpdates.flatMap(_.operations)
+    } yield
+      if (isSnapShot(operations))
+        InMemoryTableVersions.latestSnapshotTableVersion(operations)
+      else
+        InMemoryTableVersions.applyPartitionUpdates(PartitionedTableVersion(Map.empty))(operations)
+
+  def isSnapShot(operations: List[TableOperation]) = operations match {
+    case InitTable(_, isSnapshot) :: _ => isSnapshot
+    case _                             => throw new IllegalArgumentException("First operation should be InitTable")
   }
 
+  override def nextVersions(table: TableName, partitions: List[Partition]): F[Map[Partition, Version]] =
```

Comment:
This method has become even uglier than before as part of these changes, but we expect this method to disappear when we do the "UUID paths" change so we're not too worried about it...